zarr/storage.py

```diff
@@ -1091,8 +1091,9 @@ class FSStore(Store):
     Parameters
     ----------
     url : str
-        The destination to map. Should include protocol and path,
-        like "s3://bucket/root"
+        The destination to map. If no fs is provided, should include protocol
```

Comment:
Perhaps an even simpler alternative is if the input is already an FSMap? We could handle both those cases. 

Is the intent that this will be called from zarr.open* or xarray, or that calling code will instantiate this themselves before calling zarr.open(fsstore) ?

Reply:
Ok so it seems like there are three basic cases we need to cover:
1. The user wants to specify a URL (including protocol)
2. The user already has a filesystem initialized and wants to use that to create the FSStore
3. The user already has an FSMap object and want to use that to create the FSStore

I am already uncomfortable with the overloading of the constructor on this class to handle two different scenarios. If we expand it to cover 3, it will get even more messy. A better pattern would be

```python
class FSStore(Store):

    @classmethod
    def from_url(cls, url):
        # figure out fs and path
        return cls(path, fs)

    @classmethod
    def from_fsmap(cls, fsmap):
        # figure out fs and path
        return cls(path, fs)

    def __init__(self, path, fs):
        ...
```
However, this would change the API and break lots of code.

More generally, looking at this code has made me realize that the interface between zarr.storage and FSSpec is not very clean. `FSStore` reimplements and extends many of the MutableMapping methods that `FSMap` implements. `FSStore` calls both `self.map` methods and `self.fs` methods (some of them private 😬 ).

Reply:
I agree with all of that. As you have seen, this is a consequence of the evolution of the storage class in zarr itself - originally it was to be a MutableMap alone, so that could go into fsspec and be generally useful, but then stores started to gain additional capabilities. And here we are.

I am *not* uncomfortable with the class methods. I think we can have `__init__` stay the same (and not break code), but add `from_filesystem(fs, path, args)`, `from_mapper(mapper)` class methods. Either the calling code will determine which to call, or we add another helper to decide?

Reply:
Yeah, if `__init__` stayed the same then the rest of this seems pretty non-disruptive to me. I do think a classmethod approach looks nice.

Reply:
Thanks a lot for the feedback Greg! Just to be totally explicit, is adding an additional argument  to `__init__` ( as I have have done here with `fs`), while keeping the other arguments backwards compatible, considered a "disruptive change"? Or is this PR already sufficiently non-disruptive.

Reply:
Yeah, what is currently in the PR seems fine to me. It should be easy to just merge this into the v3 branch with minimal conflicts or extra work.

Reply:
Now that https://github.com/fsspec/filesystem_spec/pull/939 has gone in, it should be possible to create an FSStore from an FSMap. But let's address that in a future PR, since that functionality is still unreleased.